File: app/src/main/java/com/zofers/zofers/vvm/viewmodel/OfferCreationFirebaseViewModel.java
package com.zofers.zofers.vvm.viewmodel;

import android.net.Uri;

import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;
import com.zofers.zofers.model.Offer;
import com.zofers.zofers.service.ServiceCallback;

public class OfferCreationFirebaseViewModel extends AppViewModel {

	private Offer offer;

	public void createOffer(Uri image, ServiceCallback<Offer> callback) {
//


		offer.setUserID(getCurrentUser().getUid());
		// Create a storage reference from our app
		StorageReference storageRef = FirebaseStorage.getInstance().getReference();

		// Create a reference to "mountains.jpg"
		StorageReference imagesRef = storageRef.child("images/" + image.getLastPathSegment());


		UploadTask uploadTask = imagesRef.putFile(image);


		uploadTask.addOnFailureListener(exception -> {
		}).addOnSuccessListener(taskSnapshot -> {
			imagesRef.getDownloadUrl().addOnSuccessListener(uri -> {
				offer.setImageUrl(uri.toString());
				FirebaseFirestore.getInstance().collection("offer")
						.add(offer)
						.addOnSuccessListener(documentReference -> {
							callback.onSuccess(null); // todo change null
						})
						.addOnFailureListener(e -> callback.onFailure());


			});


		});


	}

	public Offer getOffer() {
		if (offer == null) {
			offer = new Offer();
		}
		return offer;
	}

	public void setOffer(Offer offer) {
		this.offer = offer;
	}
}

File: app/src/main/java/com/zofers/zofers/create/OfferCreationFirebaseViewModel.kt
package com.zofers.zofers.create

import android.content.Context
import android.net.Uri
import com.google.firebase.firestore.FirebaseFirestore
import com.zofers.zofers.AppViewModel
import com.zofers.zofers.model.Offer
import com.zofers.zofers.service.ServiceCallback

class OfferCreationFirebaseViewModel : AppViewModel() {

	var offer: Offer = Offer()

	fun createOffer(context: Context, image: Uri, callback: ServiceCallback<Offer>) {
		offer.userID = currentUser?.uid

		uploadImage(context, image, "images/offers/" + image.lastPathSegment!!) { uri ->
			offer.imageUrl = uri.toString()
			FirebaseFirestore.getInstance().collection("offer")
					.add(offer)
					.addOnSuccessListener {
						callback.onSuccess(null) // todo change null
					}
					.addOnFailureListener {
						callback.onFailure()
					}
		}
	}
}
